### PR TITLE
Gui display files

### PR DESCRIPTION
**Pull request by marySalvi**
Displays supported images(svs and tif) in directory list.

image

### CODE REVIEW

File: imagedephi/redact.py [hunk 1]
```diff
@@ -114,32 +114,43 @@ def get_base_rules():
...
+
 def redact_images(
     image_dir: Path,
     output_dir: Path,
     override_rules: RuleSet | None = None,
     overwrite: bool = False,
 ) -> None:
     base_rules = get_base_rules()
-    for child in image_dir.iterdir():
+    for image_file in iter_image_files(image_dir):
```
Comment (marySalvi) [inline imagedephi/redact.py:135]: @naglepuff note the refactor done here.

**Comment by manthey** [comment]
Perhaps foolishly, I tried this on a directory where I have LOTS of files (a spot where I had sub directories with ~8000 files in one directory and lots in others).  I had to search for my directory name.  I would have expected that I would drill all the way into a directory before showing files -- and then if there were a lot it wouldn't necessarily load them all.  Once I navigate all the way into a directory, it seems odd that I have to navigate back up to select the directory -- I'd expect that the CHOOSE DIRECTORIES would choose the directories I am in rather than selecting a directory and then choosing it.

**Comment by manthey** [comment]
Can we please sort the directories (at least when it starts in my root directory, the order seems arbitrary)?

**Comment by marySalvi** [comment]
> Perhaps foolishly, I tried this on a directory where I have LOTS of files (a spot where I had sub directories with ~8000 files in one directory and lots in others). I had to search for my directory name. I would have expected that I would drill all the way into a directory before showing files -- and then if there were a lot it wouldn't necessarily load them all. Once I navigate all the way into a directory, it seems odd that I have to navigate back up to select the directory -- I'd expect that the CHOOSE DIRECTORIES would choose the directories I am in rather than selecting a directory and then choosing it.

Perhaps change to click for select, double click to navigate into the directory?

**Comment by manthey** [comment]
> Perhaps change to click for select, double click to navigate into the directory?

Unless you can choose multiple directories, I think the current behavior is surprising.  When I've navigated into a directory, there isn't any way to select it -- you have to navigate back out.  Perhaps the "CHOOSE DIRECTORIES" button should be "use these directories", and it would use the directories you are in for input and output.  Selecting a directory would still enter it and show its contents; there wouldn't be any radio button, since the directory listed would be the selected directory.  For instance, in this screenshot:
image
my initial assumption was that it would use /mnt/data3/TCGA and /tmp, and if I wanted to use /mnt/data3/TCGA/square and /tmp/r, I should click each of those directories.  This would change the file list to just be the current directory (those in /mnt/data/TCGA rather than those in each of the directories in /mnt/data/TCGA), which would be more obvious to me what is about to be operated on, since it is all of the images you see rather than a subset.

File: imagedephi/redact.py [hunk 2]
```diff
@@ -114,32 +114,43 @@ def get_base_rules():
...
 
 
+def iter_image_files(directory: Path) -> Generator[Path, None, None]:
+    """
+    Given a directory return an iterable of available images.
+
+    May raise a PermissionError if the directory is not readable.
+    """
+    for child in directory.iterdir():
+        if child.suffix == ".tif" or child.suffix == ".svs":
```
Comment (manthey) [inline imagedephi/redact.py:124]: We should check if it is in `FILE_EXTENSION_MAP`.

**Comment by jeffbaumes** [comment]
> > Perhaps change to click for select, double click to navigate into the directory?
> 
> Unless you can choose multiple directories, I think the current behavior is surprising. When I've navigated into a directory, there isn't any way to select it -- you have to navigate back out. Perhaps the "CHOOSE DIRECTORIES" button should be "use these directories", and it would use the directories you are in for input and output. Selecting a directory would still enter it and show its contents; there wouldn't be any radio button, since the directory listed would be the selected directory. For instance, in this screenshot: image my initial assumption was that it would use /mnt/data3/TCGA and /tmp, and if I wanted to use /mnt/data3/TCGA/square and /tmp/r, I should click each of those directories. This would change the file list to just be the current directory (those in /mnt/data/TCGA rather than those in each of the directories in /mnt/data/TCGA), which would be more obvious to me what is about to be operated on, since it is all of the images you see rather than a subset.

I think @manthey is on the right track, where a directory with no subdirectory selected means "select the current folder". Google Drive's "move" dialog has similar behavior, where the "move" button changes name to "move here" to distinguish between these states. Since we have two folders to select that button name change doesn't make sense. Note that Google Drive also allows you to deselect a child folder by clicking on it again to essentially reselect the parent.

Feb-16-2023 09-32-12

Similarly Google Drive's folder upload opens Mac's folder selection dialog, which seems to always have an active "Upload" button whether the user has chosen a folder or not. In the latter case it's selecting the folder the user has navigated to. It does not seem the Mac dialog easily allows folder deselection (you need to command-click to do it).

We don't need to follow these exactly, but in general I feel it's good to follow a UX pattern from somewhere else when we can, kind of like deciding to use a library, we can leverage knowledge from others instead of reinventing. There may be other "folder pickers" out there to consider.

File: imagedephi/redact.py (hunk 2 above)
Comment (marySalvi) [reply imagedephi/redact.py:124]: Are there additional extensions I should include in the map or just add as we go?

Comment (manthey) [reply imagedephi/redact.py:124]: For now, we have a very limited tool that only works on tiff files and variations.  There are lots of extensions that we've seen for tiff files (ptif for Philips' generated tiff files, for instance).  From a UI perspective, the existing file extension map is file.  Eventually, we'll want to be able to let the user filter by their own method (including trying all files), and not care about file extension.

The more correct approach is to use something like the "magic" library to detect file format, or, more certain, use large_image and its tile sources -- but this requires reading some of each file and is therefore slow, especially in cases where we have thousands of files in a directory.

**Comment by marySalvi** [comment]
> Perhaps foolishly, I tried this on a directory where I have LOTS of files (a spot where I had sub directories with ~8000 files in one directory and lots in others). I had to search for my directory name. I would have expected that I would drill all the way into a directory before showing files -- and then if there were a lot it wouldn't necessarily load them all. Once I navigate all the way into a directory, it seems odd that I have to navigate back up to select the directory -- I'd expect that the CHOOSE DIRECTORIES would choose the directories I am in rather than selecting a directory and then choosing it.


fixed in 34f43a442f1b267ec0acef726e0fab93f7387bc1 and 30949e8046d2b11e2713536dbf34874e13b1849e

**Comment by marySalvi** [comment]
> Can we please sort the directories (at least when it starts in my root directory, the order seems arbitrary)?

fixed in 366b107145e10569ad4261726b5e240487e3331f

File: imagedephi/gui.py [hunk 3]
```diff
@@ -27,40 +44,25 @@ def select_directory(
     input_directory: Path = Path("/"),  # noqa: B008
     output_directory: Path = Path("/"),  # noqa: B008
 ):
+    Path("")
```
Comment (brianhelba) [inline imagedephi/gui.py:47]: @brianhelba This shouldn't be committed, remove it.

Comment (marySalvi) [reply imagedephi/gui.py]: removed in d6e4e2812808229709879d48721e30756d7fc1bd

File: imagedephi/redact.py (hunk 2 above)
Comment (marySalvi) [reply imagedephi/redact.py:124]: filed as issue #61 to fix after merge if #55  is not merged first

**Review by brianhelba — APPROVED** [review]

**Pull request merged**